File: 2. 数据结构/第4章 图论算法/3.最短路径/例4-6 香甜的黄油/butter.cpp
#include<iostream>
#include<cstdio>
#include<cmath>
#include<cstdlib>
#include<cstring>
using namespace std;
long b[1600],dis[1600],w[1600][1600],tm[1605],a[1600][1600];
bool ex[1600]={false};
int main ()
{
	long long n,p,c,i,j,x,y,ma,head,tail,s=0,u;
	cin>>n>>p>>c; 
	for (i=1; i<=n; i++)
	  cin>>b[i];
	for(i=1; i<=c; i++)
	{
		cin>>x>>y; cin>>w[x][y];
		a[x][0]++;
		a [x][a[x][0]]=y;
		a[y][0]++;
		a[y][a[y][0]]=x;
		w[y][x]=w[x][y];
	}
	ma=1000001;
	for (i=1; i<=p; i++)
	{
	{
		for(j=1; j<=c; j++)dis[j]=1000001;
	    for (j=1; j<=p; j++)ex[j]=false; 
	    dis[i]=0; tm[1]=i; head=0; tail=1; ex[i]=true; s=0;
		do
		{
			head=head%1061+1;
			u=tm[head];  ex[u]=false;
			for (j=1; j<=a[u][0];  j++)
			if (dis[a[u][j]]>dis[u]+w[u][a[u][j]])
			{
			dis[a[u][j]]=dis[u]+w[u][a[u][j]];
			if (ex[a[u][j]]==false)
			{
				tail=tail%1061+1;
				tm[tail]=a[u][j];
				ex[a[u][j]]=true;
			}
			}
		}while (head!=tail);
	}
	for (j=1; j<=n; j++)
	{
	s=s+dis[b[j]];
	}
	if (s<ma) ma=s;}
	cout<<ma;
}

File: 2. 数据结构/第4章 图论算法/3.最短路径/例4-6 香甜的黄油/butter.pas
program LJ;
var n,p,c,i,j,x,y,min,head,tail,tot,u:longint;
    dis,b,num:array[1..800]of longint;
    w:array[1..800,1..800]of longint;
    team:array[1..1605] of longint;
a:array[1..800,0..800]of longint;
exist:array[1..800] of boolean;
begin
  assign(input,'butter.in');reset(input);
  assign(output,'butter.out');rewrite(output);
  readln(n,p,c);
  for i:=1 to n do readln(b[i]);
  for i:=1 to p do
    for j:=1 to p do
      w[i,j]:=maxlongint div 3;
  for i:=1 to c do
  begin
    readln(x,y,w[x,y]);
    inc(num[x]);
    a[x,num[x]]:=y;
    inc(num[y]);
    a[y,num[y]]:=x;
    w[y,x]:=w[x,y];
  end;
  min:=maxlongint;
  for i:=1 to p do
  begin
    for j:=1 to p do dis[j]:=maxlongint div 3;
    fillchar(team,sizeof(team),0);
    fillchar(exist,sizeof(exist),false);
    dis[i]:=0;team[1]:=i;head:=0;tail:=1; exist[i]:=true;
    repeat
      inc(head);
      head:=((head-1) mod 1601)+1;
      u:=team[head]; exist[u]:=false;
      for j:=1 to num[u] do
        if dis[a[u,j]]>dis[u]+w[u,a[u,j]] then
        begin
          dis[a[u,j]]:=dis[u]+w[u,a[u,j]];
          if exist[ a[u,j] ]=false then
           begin
             inc(tail); tail:=((tail-1) mod 1601)+1;
             team[tail]:=a[u,j];
             exist[ a[u,j] ]:=true;
           end;
      end;
    until head=tail;
    tot:=0;
    for j:=1 to n do tot:=tot+dis[b[j]];
    if tot<min then min:=tot;
  end;
  writeln(min);
  close(input);close(output);
end.
